File: LeetCodeTest/20.有效的括号.java
/*
 * @lc app=leetcode.cn id=20 lang=java
 *
 * [20] 有效的括号
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        if(s.length()%2!=0) return false;
        Map<Character,Character> map = new HashMap<>();
        map.put('}','{');
        map.put(')','(');
        map.put(']','[');
        Stack<Character> stack = new Stack<>();
        Deque<Character> deque = new LinkedList<>();
        deque.isEmpty();
        for(int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            if(map.containsKey(c)){
                if(stack.isEmpty() || stack.peek() != map.get(c)){
                    return false;
                }
                stack.pop();
            } else {
                stack.push(c);
            }
        }
        return stack.isEmpty();
    }
}
// @lc code=end

File: LeetCodeTest/20.有效的括号.cpp
/*
 * @lc app=leetcode.cn id=20 lang=cpp
 *
 * [20] 有效的括号
 */

// @lc code=start
class Solution {
public:
    bool isValid(string s) {
        if(s.length() % 2 != 0) return false;
        unordered_map<char,char> map = {
            {'}','{'},
            {']','['},
            {')','('}
        };
        stack<char> stack;
        for(char c:s){
            if(map.count(c)>0){
                if(stack.empty() || map[c]!=stack.top()){
                    return false;
                }
                stack.pop();
            } else {
                stack.push(c);
            }
        
        }
        return stack.empty();
    }
};
// @lc code=end

